custom/icds_reports/static/js/angular-services/locations.service.js

```diff
@@ -122,19 +131,191 @@ window.angular.module('icdsApp').factory('locationsService', ['$http', '$locatio
+                        var childSelected = _.find(locations, function(location) {
+                            return location.location_id === childSelectedId;
+                        });
+                        vm.selectedLocations[levelOfSelectedLocation] = childSelected.parent_id;
+                        levelOfSelectedLocation -= 1;
+                    }
+
+                    var levels = [];
+                    window.angular.forEach(vm.levels, function (value) {
+                        if (value.id > selectedLocationIndex()) {
+                            levels.push(value);
+                        }
+                    });
+                    vm.groupByLevels = levels;
+                    vm.selectedLevel = selectedLocationIndex() + 1;
+                });
+            } else {
+                vm.locationPromise = this.getRootLocations().then(function(data) {
+                    locationsCache.root = [NATIONAL_OPTION].concat(data.locations);
+                });
+                vm.groupByLevels = vm.levels;
+            }
+            return locationsCache;
+        },
+
+        onSelectLocation : function(item, level, locationsCache, vm) {
+            this.resetLevelsBelow(level, vm);
+            if (level < 4) {
+                vm.locationPromise = this.getChildren(item.location_id).then(function (data) {
+                    if (item.user_have_access) {
+                        locationsCache[item.location_id] = [ALL_OPTION].concat(data.locations);
+                        vm.selectedLocations[level + 1] = ALL_OPTION.location_id;
+                    } else {
+                        locationsCache[$item.location_id] = data.locations;
+                        vm.selectedLocations[level + 1] = data.locations[0].location_id;
+                        vm.onSelect(data.locations[0], level + 1);
+                    }
+                });
+            }
+            var selectedLocationIndex = this.selectedLocationIndex(vm.selectedLocations);
+            vm.selectedLocationId = vm.selectedLocations[selectedLocationIndex];
+            var levels = [];
+            vm.selectedLevel = selectedLocationIndex + 1;
+            window.angular.forEach(vm.levels, function (value) {
+                if (value.id > selectedLocationIndex) {
+                    levels.push(value);
+                }
+            });
+            vm.groupByLevels = levels;
+        },
+
+        getLocationPlaceholder : function(locationTypes, disallowNational) {
+            return _.map(locationTypes, function(locationType) {
+                if (locationType.name === 'state') {
+                    if (disallowNational) {
+                        return 'Select State';
+                    } else {
+                        return NATIONAL_OPTION.name;
+                    }
+                }
+                return locationType.name;
+            }).join(', ');
+        },
+
+        isLocationDisabled : function(level, vm) {
```

Comment:
Nitpick and this is purely pattern matching, but it seems like most functions that take `vm` make it the first param.

Reply:
(after reading more of the code, maybe this isn't true)